# Statistical Analysis of High-Throughput Genomic and Transcriptomic Data
Fall/Herbst-semester 2024

## Lectures
Mondays 9.00-9.45 (Y27-H-46), 10.00-10.45 (Y27-H-46)

## Exercises
Monday 11.00-11.45

## Lecturers

[Dr. Hubert Rehrauer](http://www.fgcz.ch/the-center/people/rehrauer.html), Group Leader of Genome Informatics at [FGCZ](http://www.fgcz.ch/)

[Prof. Dr. Mark Robinson](https://robinsonlabuzh.github.io/), Professor of Statistical Genomics, [DMLS](https://www.mls.uzh.ch/en.html), UZH

[Dr. Pierre-Luc Germain](https://schrattlab.ethz.ch/the-group/people/person-detail.MjUwODk0.TGlzdC8yOTEzLDExMTE5MzUxNzY=.html), Lecturer, [D-HEST](https://hest.ethz.ch/en), ETHZ

## Schedule

| Date  | Lecturer | Topic | Exercise | JC1 | JC2 |
| --- | --- | --- | --- | --- | --- |
| 16.09.2024  | Mark | admin; mol. bio. basics | quarto; git(hub) | | |
| 23.09.2024  | Mark | interactive technology/statistics session  | group exercise: technology pull request | | |
| 30.09.2024  | Mark  | limma + friends | linear model simulation + design matrices | | |
| 07.10.2024  | Hubert | NGS intro; exploratory data analysis | EDA in R | | |
| 14.10.2024  | Hubert | mapping  | Rsubread | | |
| 21.10.2024  | Hubert | RNA-seq quantification | RSEM  | X | X | 
| 28.10.2024  | Mark | edgeR+friends 1 | basic edgeR/voom | [Validation of noise models for single-cell transcriptomics](https://doi.org/10.1038/nmeth.2930) (GC, YD) | X |
| 04.11.2024  | Mark | edgeR+friends 2 | advanced edgeR/voom | X | [Robust identification of perturbed cell types in single-cell RNA-seq data](https://www.nature.com/articles/s41467-024-51649-3)(AL, JM) |
| 11.11.2024  | Mark | hands-on session #1: RNA-seq |  FASTQC/Salmon/etc. | X | [MAST: a flexible statistical framework for assessing transcriptional changes and characterizing heterogeneity in single-cell RNA sequencing data](https://doi.org/10.1186/s13059-015-0844-5) (DH) |
| 18.11.2024  | Hubert | single-cell 1: preprocessing, dim. reduction, clustering | clustering | [MOFA+: a statistical framework for comprehensive integration of multi-modal single-cell data - Genome Biology](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1) (ZL, PM) | [Identifying genetic variants that influence the abundance of cell states in single-cell data](https://www.nature.com/articles/s41588-024-01909-1) (JF, TH) |
| 25.11.2024  | Mark | hands-on session #2: cytometry | cytof null comparison | [Significance analysis for clustering with single-cell RNA-sequencing data](https://www.nature.com/articles/s41592-023-01933-9) (HS, ZM) | [Genomic prediction in multi-environment trials in maize using statistical and machine learning methods](https://www.proquest.com/docview/2913316100/4BBE80EEBE564D48PQ/3?sourcetype=Scholarly%20Journals) (JB)|
| 02.12.2024  | Mark | single-cell 2: clustering, marker gene DE  | marker gene DE | [Quantifying cell-state densities in single-cell phenotypic landscapes using Mellon](https://www.nature.com/articles/s41592-024-02302-w) (MW) | [Inference and analysis of cell-cell communication using CellChat](https://doi.org/10.1038/s41467-021-21246-9) (L&M)|
| 09.12.2024  | Pierre-Luc | hands-on session #3: single-cell RNA-seq (cell type definition, differential state)  | full scRNA-seq pipeline | [IDEAS: individual level differential expression analysis for single-cell RNA-seq data](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02605-1) (GO, JV) | [Powerful and accurate detection of temporal gene expression patterns from multi-sample multi-stage single-cell transcriptomics data with TDEseq](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-024-03237-3) (AVK)  |
| 16.12.2024  | Mark | spatial omics  | spatial statistics | X | X |


## Course material

Assuming you have git installed locally, you can check out the entire set of course materials with the following command (from command line):
```
git clone https://github.com/sta426hs2024/material.git
```
and get updates by running `git pull` at any later time in the same directory.

Alternatively, to retrieve a ZIP file of the repository, you can click on the (green) 'Code' button (top right of main panel) and then click 'Download ZIP'.

